PIP_COMPATIBILITY.md

```diff
@@ -0,0 +1,317 @@
+# requirements.txt
+git+https://github.com/pallets/flask
+```
+
+This is a common pattern in `pip` workflows, and is used to install a package from a direct URL
+without incorporating the package name upfront.
+
+`uv` supports direct URL dependencies from HTTP and VCS sources, but requires that the package name
+be provided upfront, as in `uv install "flask @ git+https://github.com/pallets/flask"`, or:
+
+```txt
+# requirements.txt
+flask @ git+https://github.com/pallets/flask
+```
+
+In the future, `uv` will support direct URL dependencies without package names. For more, see
+[#313](https://github.com/astral-sh/uv/issues/313).
+
+## Transitive direct URL dependencies
+
+While uv does support direct URL dependencies (e.g., `black @ https://...`), it does not support
+nested (or "transitive") direct URL dependencies, instead requiring that any direct URL dependencies
+are declared upfront.
+
+For example, if `black @ https://...` itself had a dependency on `toml @ https://...`, uv would
+reject the transitive direct URL dependency on `toml` and require that `toml` be declared as a
+dependency in the `pyproject.toml` file, like:
+
+```toml
+# pyproject.toml
+dependencies = [
+    "black @ https://...",
+    "toml @ https://...",
+]
+```
+
+This is a deliberate choice to avoid the correctness and security issues associated with allowing
+transitive dependencies to introduce arbitrary URLs into the dependency graph.
+
+For example:
+
+- Your package depends on `package_a==1.0.0`.
+- Your package depends on `package_b==1.0.0`.
+- `package_b==1.0.0` depends on `package_a @ https://...`.
+
+If `package_a @ https://...` happens to resolve to version `1.0.0`, `pip` would install `package_a`
+from the direct URL. This is a security issue, since the direct URL could be controlled by an
+attacker, and a correctness issue, since the direct URL could resolve to an entirely different
+package with the same name and version.
+
+In the future, uv may allow transitive URL dependencies in some form (e.g., with user opt-in).
+For more, see [#1808](https://github.com/astral-sh/uv/issues/1808).
+
+## Pre-release compatibility
+
+By default, uv will accept pre-release versions during dependency resolution in two cases:
+
+1. If the package is a direct dependency, and its version markers include a pre-release specifier
+   (e.g., `flask>=2.0.0rc1`).
+1. If _all_ published versions of a package are pre-releases.
+
+If dependency resolution fails due to a transitive pre-release, uv will prompt the user to
+re-run with `--prerelease=allow`, to allow pre-releases for all dependencies.
+
+Alternatively, you can add the transitive dependency to your `requirements.in` file with
+pre-release specifier (e.g., `flask>=2.0.0rc1`) to opt in to pre-release support for that specific
+dependency.
+
+In sum, uv needs to know upfront whether the resolver should accept pre-releases for a given
+package. `pip`, meanwhile, will generally respect pre-release identifiers in transitive
```

Comment:
FYI, I expanded on this a bit here: https://github.com/astral-sh/uv/issues/1641#issuecomment-1981402429

It's not important for uv, but certainly anything stronger than "generally" would be wrong, could even be "sometimes".

Reply:
Thanks!